### PR TITLE
Adding a sample for LightGbm ranking

### LINKED ISSUE #776: Provide an example of LightgbmRanker?
@codemzs

### LINKED ISSUE #2530: Create Samples for LightGbm Ranker.
Need doc samples for lightgbm ranker. 
Reference the samples in the xml docs for the two public lightgbm ranker methods.

https://github.com/dotnet/machinelearning/blob/f8e58dd4327f4c5013abbd7f691d835bdd377909/src/Microsoft.ML.LightGBM/LightGbmCatalog.cs#L104

### PR DESCRIPTION
**Pull request by najeeb-kazmi**
Replacing PR #2650 as I messed up commit history there.

Fixes #2530
Fixes #776

- Adds a sample for LightGbm ranking. 
- Cleans up namespaces in Microsoft.ML.Samples project.
- Addresses feedback from first round of #2650.

### CODE REVIEW

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Ranking/LightGBMRankingWithOptions.cs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+            var split = mlContext.Ranking.TrainTestSplit(dataview, testFraction: 0.1, samplingKeyColumn: "GroupId");
+
+            // Create the Estimator pipeline. For simplicity, we will train a small tree with 4 leaves and 2 boosting iterations.
+            var pipeline = mlContext.Ranking.Trainers.LightGbm(
+                new Options
+                {
+                    NumLeaves = 4,
+                    MinDataPerLeaf = 10,
+                    LearningRate = 0.1,
+                    NumBoostRound = 2
```
Comment (Ivanidzo4ka) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Ranking/LightGBMRankingWithOptions.cs]: >NumBoostRound [](start = 20, length = 13)

please include example for how to change booster.

File: src/Microsoft.ML.SamplesUtils/SamplesDatasetUtils.cs [hunk 2]
```diff
@@ -138,14 +138,51 @@ public static IDataView LoadFeaturizedAdultDataset(MLContext mlContext)
...
             var data = reader.Read(dataFile);
             var featurizedData = pipeline.Fit(data).Transform(data);
             return featurizedData;
         }
 
+        public static string DownloadMslrWeb10k()
+        {
+            var fileName = "MSLRWeb10KTrain720kRows.tsv";
+            if (!File.Exists(fileName))
+                Download("https://tlcresources.blob.core.windows.net/datasets/MSLR-WEB10K/MSLR-WEB10K_Fold1.TRAIN.500MB_720k-rows.tsv", fileName);
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.SamplesUtils/SamplesDatasetUtils.cs]: >https://tlcresources.blob.core.windows.net/datasets/MSLR-WEB10K/MSLR-WEB10K_Fold1.TRAIN.500MB_720k-rows.tsv [](start = 26, length = 107)

This thing is 500 Mb, why we using it?

File: src/Microsoft.ML.SamplesUtils/SamplesDatasetUtils.cs [hunk 3]
```diff
@@ -138,14 +138,51 @@ public static IDataView LoadFeaturizedAdultDataset(MLContext mlContext)
...
+                    new TextLoader.Column("GroupId", DataKind.TX, 1),
+                    new TextLoader.Column("Features", DataKind.R4, new[] { new TextLoader.Range(2, 138) })
+                }
+            );
+
+            // Load the raw dataset.
+            var data = reader.Read(dataFile);
+
+            // Create the featurization pipeline. First, hash the GroupId column.
+            var pipeline = mlContext.Transforms.Conversion.Hash("GroupId")
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.SamplesUtils/SamplesDatasetUtils.cs]: >mlContext.Transforms.Conversion.Hash("GroupId") [](start = 27, length = 47)

Doesn't TrainTestSplit take care of that?

File: src/Microsoft.ML.SamplesUtils/SamplesDatasetUtils.cs (hunk 2 above)
Comment (najeeb-kazmi) [reply src/Microsoft.ML.SamplesUtils/SamplesDatasetUtils.cs]: @Ivanidzo4ka This is the only actual ranking dataset we have. I know we use Adult dataset to do some ranking tests with a fake group id column, but I'm not going to add it to our ranking samples. It will only invite confused questions as to why we are using column X as group id, why are we doing ranking on a binary classification dataset, etc. I think it's best to avoid all of that and show an actual ranking example. 

I can either switch to using the validation set for this example, which is 160MB, or make a smaller subset of the 500MB dataset. What do you think?

Comment (Ivanidzo4ka) [reply src/Microsoft.ML.SamplesUtils/SamplesDatasetUtils.cs]: Can we just generate one?
160 MB is too much.
you generated dataset in which you have 10 features, you have 10 different groups, and label depends on value in feature_x if x is value in group column.
Or since you do example for lightgbm you can use [https://github.com/Microsoft/LightGBM/tree/master/examples/lambdarank](https://github.com/Microsoft/LightGBM/tree/master/examples/lambdarank) ligthgbm dataset

---
In reply to: [260045081](https://github.com/dotnet/machinelearning/pull/2704#discussion_r260045081) [](ancestors = 260045081)

Comment (najeeb-kazmi) [reply src/Microsoft.ML.SamplesUtils/SamplesDatasetUtils.cs]: Hmm that's not a very easy to follow example. There is no group id in the data, nor does the .conf file specify anything about it. It's very unclear how that is being handled or whether the ranker is being trained with one row per group (and if so, why). I'll just make a small sample of MSLR dataset, something like 10,000 rows, ~7MB.

Comment (Ivanidzo4ka) [reply src/Microsoft.ML.SamplesUtils/SamplesDatasetUtils.cs]: It specifies query file. in which you have list of numbers. take first number, it's gonna be amount of examples for first group. take second number -> number of examples for second group and so on.
same for test.
My main concern was regarding publicity of that MSLR dataset, but it looks like it become public quite a while ago [https://www.microsoft.com/en-us/research/project/mslr/](https://www.microsoft.com/en-us/research/project/mslr/)


---
In reply to: [260058524](https://github.com/dotnet/machinelearning/pull/2704#discussion_r260058524) [](ancestors = 260058524)

Comment (najeeb-kazmi) [reply src/Microsoft.ML.SamplesUtils/SamplesDatasetUtils.cs]: Yes, and we also have approval to use this in our samples with attribution information in the [README] (https://github.com/dotnet/machinelearning/blob/master/test/data/README.md).

File: src/Microsoft.ML.SamplesUtils/SamplesDatasetUtils.cs (hunk 3 above)
Comment (najeeb-kazmi) [reply src/Microsoft.ML.SamplesUtils/SamplesDatasetUtils.cs]: No, GroupId needs to be converted to Key first.

Comment (Ivanidzo4ka) [reply src/Microsoft.ML.SamplesUtils/SamplesDatasetUtils.cs]: https://github.com/dotnet/machinelearning/blob/3b9d407d9dc4f8c46fa85ab80575ef16d74df6df/src/Microsoft.ML.Data/TrainCatalog.cs#L201

Are you sure?

Comment (najeeb-kazmi) [reply src/Microsoft.ML.SamplesUtils/SamplesDatasetUtils.cs]: I mean, I get that error when I try this pipeline without hashing GroupId
